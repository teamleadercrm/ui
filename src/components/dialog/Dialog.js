import React, { createRef, PureComponent } from 'react';
import PropTypes from 'prop-types';
import omit from 'lodash.omit';
import cx from 'classnames';

import theme from './theme.css';

import { Button, ButtonGroup, DialogBase, Heading3 } from '../../index';
import { IconDragMediumFilled } from '@teamleader/ui-icons';

class Dialog extends PureComponent {
  bodyRef = createRef();
  dragHandleRef = createRef();

  getHeader = () => {
    const { onCloseClick, title } = this.props;

    const dragIcon = (
      <div className={theme['drag-icon']} ref={this.dragHandleRef}>
        <IconDragMediumFilled />
      </div>
    );

    return (
      <DialogBase.Header icon={dragIcon} onCloseClick={onCloseClick}>
        <Heading3 maxLines={1}>{title}</Heading3>
      </DialogBase.Header>
    );
  };

  getFooter = () => {
    const { tertiaryAction, secondaryAction, primaryAction, leftAction } = this.props;

    return (
      <DialogBase.Footer display="flex" justifyContent={leftAction ? 'space-between' : 'flex-end'}>
        {leftAction && <Button {...leftAction} />}
        <ButtonGroup justifyContent="flex-end">
          {tertiaryAction && <Button {...tertiaryAction} level="link" />}
          {secondaryAction && <Button {...secondaryAction} />}
          <Button level="primary" {...primaryAction} />
        </ButtonGroup>
      </DialogBase.Footer>
    );
  };

  render() {
    const { children, className, scrollable, title, ...otherProps } = this.props;

    const classNames = cx(theme['dialog'], className);

    const restProps = omit(otherProps, [
      'onCloseClick',
      'primaryAction',
      'secondaryAction',
      'tertiaryAction',
      'leftAction',
    ]);

    return (
      <DialogBase
        className={classNames}
        {...restProps}
        scrollable={false}
        initialFocusRef={this.bodyRef}
        dragHandleRef={this.dragHandleRef}
      >
        {title && this.getHeader()}
        <DialogBase.Body ref={this.bodyRef} scrollable={scrollable}>
          {children}
        </DialogBase.Body>
        {this.getFooter()}
      </DialogBase>
    );
  }
}

Dialog.propTypes = {
  /** If true, the dialog will show on screen. */
  active: PropTypes.bool,
  /** The content to display inside the dialog. */
  children: PropTypes.any,
  /** A class name for the wrapper to apply custom styles. */
  className: PropTypes.string,
  /** Object containing the the props of the action on the left (a Button). */
  leftAction: PropTypes.object,
  /** Callback function that is fired when the close icon (in the header) is clicked. */
  onCloseClick: PropTypes.func,
  /** Object containing the props of the primary action (a Button, with level prop set to 'primary'). */
  primaryAction: PropTypes.object.isRequired,
  /** The size of the dialog. */
  size: PropTypes.oneOf(['small', 'medium', 'large', 'fullscreen']),
  /** If true, the content of the dialog will be scrollable when it exceeds the available height. */
  scrollable: PropTypes.bool,
  /** Object containing the the props of the secondary action (a Button). */
  secondaryAction: PropTypes.object,
  /** Object containing the props of the tertiary action (a Link, with the inherit props set to false). */
  tertiaryAction: PropTypes.object,
  /** The title of the dialog. */
  title: PropTypes.string,
};

Dialog.defaultProps = {
  scrollable: true,
  size: 'medium',
};

export default Dialog;
