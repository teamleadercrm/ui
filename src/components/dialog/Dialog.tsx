import { IconDragMediumFilled } from '@teamleader/ui-icons';
import cx from 'classnames';
import omit from 'lodash.omit';
import React, { ReactNode, useRef } from 'react';
import { Button, ButtonGroup, DialogBase, Heading3 } from '../../index';
import { DialogBaseProps } from './DialogBase';
import theme from './theme.css';

interface DialogProps extends Omit<DialogBaseProps, 'ref'> {
  /** If true, the dialog will show on screen. */
  active?: boolean;
  /** The content to display inside the dialog. */
  children?: ReactNode;
  /** A class name for the wrapper to apply custom styles. */
  className?: string;
  /** Object containing the the props of the action on the left (a Button). */
  leftAction?: object;
  /** Callback function that is fired when the close icon (in the header) is clicked. */
  onCloseClick?: () => void;
  /** Object containing the props of the primary action (a Button, with level prop set to 'primary'). */
  primaryAction: object;
  /** The size of the dialog. */
  size?: 'small' | 'medium' | 'large' | 'fullscreen';
  /** If true, the content of the dialog will be scrollable when it exceeds the available height. */
  scrollable?: boolean;
  /** Object containing the the props of the secondary action (a Button). */
  secondaryAction?: object;
  /** Object containing the props of the tertiary action (a Link, with the inherit props set to false). */
  tertiaryAction?: object;
  /** The title of the dialog. */
  title?: string;
  /** If true the dialog will render as a form element. */
  form?: boolean;
  /** Optional callback if the dialog is a form and is being submitted */
  onSubmit?: () => void;
}

const Dialog = ({
  children,
  className,
  scrollable = true,
  title,
  form,
  onSubmit,
  tertiaryAction,
  secondaryAction,
  primaryAction,
  leftAction,
  onCloseClick,
  ...otherProps
}: DialogProps) => {
  const bodyRef = useRef<HTMLElement>(null);
  const dragHandleRef = useRef<HTMLDivElement>(null);

  const getHeader = () => {
    const dragIcon = (
      <div className={theme['drag-icon']} ref={dragHandleRef}>
        <IconDragMediumFilled />
      </div>
    );

    return (
      <DialogBase.Header icon={dragIcon} onCloseClick={onCloseClick}>
        <Heading3 maxLines={1}>{title}</Heading3>
      </DialogBase.Header>
    );
  };

  const getFooter = () => {
    return (
      <DialogBase.Footer display="flex" justifyContent={leftAction ? 'space-between' : 'flex-end'}>
        {leftAction && <Button {...leftAction} />}
        <ButtonGroup justifyContent="flex-end">
          {tertiaryAction && <Button {...tertiaryAction} level="link" />}
          {secondaryAction && <Button {...secondaryAction} />}
          <Button level="primary" {...primaryAction} />
        </ButtonGroup>
      </DialogBase.Footer>
    );
  };

  const classNames = cx(theme['dialog'], className);

  const restProps = omit(otherProps, [
    'onCloseClick',
    'primaryAction',
    'secondaryAction',
    'tertiaryAction',
    'leftAction',
  ]);

  return (
    <DialogBase
      className={classNames}
      {...restProps}
      scrollable={false}
      initialFocusRef={bodyRef}
      dragHandleRef={dragHandleRef}
      form={form}
      onSubmit={onSubmit}
    >
      {title && getHeader()}
      <DialogBase.Body ref={bodyRef} scrollable={scrollable}>
        {children}
      </DialogBase.Body>
      {getFooter()}
    </DialogBase>
  );
};

export default Dialog;
