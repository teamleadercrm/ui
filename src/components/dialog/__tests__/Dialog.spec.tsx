import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Dialog from '../Dialog';

describe('Component - Dialog', () => {
  it('renders', async () => {
    const screen = render(
      <Dialog
        active
        title="Foobar"
        primaryAction={{
          label: 'Confirm',
        }}
        secondaryAction={{
          label: 'Cancel',
        }}
        tertiaryAction={{
          children: 'Delete',
        }}
        onCloseClick={() => {}}
        onOverlayClick={() => {}}
        onEscKeyDown={() => {}}
      >
        Foobar
      </Dialog>,
    );
    expect(screen.baseElement).toMatchSnapshot();
  });

  it('does not render anything when the dialog is not active', async () => {
    const screen = render(
      <Dialog
        active={false}
        title="Foobar"
        primaryAction={{
          label: 'Confirm',
        }}
      >
        Foobar
      </Dialog>,
    );
    expect(screen.baseElement).toMatchSnapshot();
  });

  it('calls the onClose handler when clicking the close button', async () => {
    const handleOnClose = jest.fn();
    const user = userEvent.setup();

    const screen = render(
      <Dialog
        active
        title="Foobar"
        primaryAction={{
          label: 'Confirm',
        }}
        onCloseClick={handleOnClose}
      >
        Foobar
      </Dialog>,
    );

    expect(handleOnClose).not.toBeCalled();
    const closeButton = screen.baseElement.querySelector('section button[type="button"]');

    await user.click(closeButton!);
    expect(handleOnClose).toBeCalledTimes(1);
  });

  it('calls the onClose handler when clicking the overlay', async () => {
    const handleOnClose = jest.fn();
    const user = userEvent.setup();

    const screen = render(
      <Dialog
        active
        title="Foobar"
        primaryAction={{
          label: 'Confirm',
        }}
        onOverlayClick={handleOnClose}
      >
        Foobar
      </Dialog>,
    );

    expect(handleOnClose).not.toBeCalled();
    const overlay = screen.baseElement.querySelector('[data-teamleader-ui="overlay"]');

    await user.click(overlay!);
    expect(handleOnClose).toBeCalledTimes(1);
  });

  it('renders a form', async () => {
    const screen = render(
      <Dialog
        active
        title="Foobar"
        primaryAction={{
          label: 'Confirm',
        }}
        form
      >
        Foobar
      </Dialog>,
    );

    expect(screen.baseElement.querySelector('form')).toBeVisible();
  });

  it('calls the form onSubmit when pressing the submit button', async () => {
    const handleSubmit = jest.fn((event) => event.preventDefault()); // need to add this preventDefault, otherwise it errors
    const user = userEvent.setup();

    const screen = render(
      <Dialog
        active
        title="Foobar"
        primaryAction={{
          label: 'Confirm',
          type: 'submit',
        }}
        form
        onSubmit={handleSubmit}
      >
        Foobar
      </Dialog>,
    );

    const submitButton = screen.getByRole('button', { name: 'Confirm' });
    await user.click(submitButton);

    expect(handleSubmit).toBeCalled();
  });
});
