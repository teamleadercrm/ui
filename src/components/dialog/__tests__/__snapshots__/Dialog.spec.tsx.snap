// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - Dialog it does not render anything when the dialog is not active 1`] = `
<body
  style="overflow: hidden;"
>
  <div />
</body>
`;

exports[`Component - Dialog it renders 1`] = `
<body
  style="overflow: hidden;"
>
  <div />
  <div
    class="overlay dark overlay"
    data-teamleader-ui="overlay"
  >
    <div>
      <div>
        <div
          tabindex="0"
        />
        <div
          class="box dialog-base is-medium is-entering dialog"
          data-teamleader-ui="dialog"
        >
          <div
            class="inner"
          >
            <section
              class="box align-items-center display-flex padding-bottom-4 padding-left-4 padding-right-4 padding-top-4 section neutral"
              data-teamleader-ui="section"
            >
              <span
                class="box align-items-center display-flex margin-right-3 teal darkest"
                data-teamleader-ui="icon"
              >
                <div
                  class="drag-icon"
                  opacity="0.84"
                >
                  <svg
                    fill="currentColor"
                    height="24"
                    preserveAspectRatio="xMidYMid meet"
                    style="opacity: 0.84; vertical-align: middle;"
                    viewBox="0 0 24 24"
                    width="24"
                  >
                    <path
                      d="M8.5 6a1.5 1.5 0 100-3 1.5 1.5 0 000 3zm0 15a1.5 1.5 0 100-3 1.5 1.5 0 000 3zm1.5-9a1.5 1.5 0 11-3 0 1.5 1.5 0 013 0zm5.5 1.5a1.5 1.5 0 100-3 1.5 1.5 0 000 3zm1.5-9a1.5 1.5 0 11-3 0 1.5 1.5 0 013 0zM15.5 21a1.5 1.5 0 100-3 1.5 1.5 0 000 3z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
              </span>
              <h3
                class="box heading heading-3 darkest overflow-singleline"
                data-teamleader-ui="heading"
              >
                Foobar
              </h3>
              <button
                class="box margin-bottom-negative-1 margin-top-negative-1 button-base icon-button is-medium close-icon"
                data-teamleader-ui="icon-button"
                type="button"
              >
                <span
                  class="box align-items-center display-flex neutral darkest"
                  data-teamleader-ui="icon"
                >
                  <svg
                    fill="currentColor"
                    height="24"
                    preserveAspectRatio="xMidYMid meet"
                    style="opacity: 0.84; vertical-align: middle;"
                    viewBox="0 0 24 24"
                    width="24"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M13.414 12l4.95-4.95a1 1 0 00-1.414-1.414L12 10.586l-4.95-4.95A1 1 0 005.636 7.05l4.95 4.95-4.95 4.95a1 1 0 001.414 1.414l4.95-4.95 4.95 4.95a1 1 0 001.414-1.414L13.414 12z"
                      fill="currentColor"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </button>
            </section>
            <div
              class="box display-flex flex-direction-column"
              style="overflow-y: auto;"
            >
              Foobar
            </div>
            <div
              class="box display-flex justify-content-flex-end padding-bottom-4 padding-left-4 padding-right-4 padding-top-4"
              style="border-top: 1px solid #e4e4e6;"
            >
              <div
                class="box justify-content-flex-end group"
                data-teamleader-ui="button-group"
              >
                <button
                  class="box reset-box-sizing reset-font-smoothing button-base button link medium"
                  data-teamleader-ui="button"
                  type="button"
                >
                  <span
                    class="box ui-text ui-text-body darkest overflow-singleline"
                    data-teamleader-ui="ui-text"
                  >
                    Delete
                  </span>
                </button>
                <button
                  class="box reset-box-sizing reset-font-smoothing button-base button secondary medium"
                  data-teamleader-ui="button"
                  type="button"
                >
                  <span
                    class="box ui-text ui-text-body darkest overflow-singleline"
                    data-teamleader-ui="ui-text"
                  >
                    Cancel
                  </span>
                </button>
                <button
                  class="box reset-box-sizing reset-font-smoothing button-base button primary medium"
                  data-teamleader-ui="button"
                  type="button"
                >
                  <span
                    class="box ui-text ui-text-body darkest overflow-singleline"
                    data-teamleader-ui="ui-text"
                  >
                    Confirm
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          tabindex="0"
        />
      </div>
    </div>
  </div>
</body>
`;
